File: frontend/src/shared/api/services/device-data.ts
import type { PaginatedResponse } from "~/shared/lib/api";

import { AxiosRequestConfig } from "axios";
import { client } from "../client";
import { Device } from "./device";

const ENDPOINT = "api/device-data/";

export enum WeatherType {
  CLEAR = "clear",
  CLOUDS = "clouds",
  RAIN = "rain",
  FOG = "fog",
}

export type DeviceData = {
  id: number;
  timestamp: string;
  device: Device;
  car_count: number;
  traffic_speed: number;
  traffic_density: number;
  pedestrian_count: number;
  pedestrian_density: number;
  ambient_light: number;
  dimming_level: number;
  lamp_power: number;
  weather: WeatherType;
};

export const getDeviceDataList = (config?: AxiosRequestConfig) => {
  return client.get<PaginatedResponse<DeviceData>>(ENDPOINT, config);
};

File: control-system/models/device_data.py
from datetime import UTC, datetime

from sqlalchemy import Column, DateTime, Float, ForeignKey, Integer, String
from sqlalchemy.orm import relationship

from core.db import Base


class DeviceDataModel(Base):
    __tablename__ = "device_data"

    id = Column(Integer, primary_key=True, index=True, autoincrement=True)
    timestamp = Column(DateTime, default=datetime.now(tz=UTC))
    device_id = Column(Integer, ForeignKey("device.id"))
    car_count = Column(Integer)
    traffic_speed = Column(Float)
    traffic_density = Column(Float)
    pedestrian_count = Column(Integer)
    pedestrian_density = Column(Float)
    ambient_light = Column(Float)
    dimming_level = Column(Float)
    lamp_power = Column(Float)
    weather = Column(String)


    device = relationship("DeviceModel", back_populates="data")
    calculated_dimming_level = relationship("DeviceDataCalculatedDimModel", back_populates="data", uselist=False)
    corrected_dimming_level = relationship("DeviceDataCorrectedDimModel", back_populates="data", uselist=False)
